memory-viz/src/memory_model.ts

```diff
@@ -1376,7 +1480,309 @@ export class MemoryModel {
+                (item.width === undefined || item.height === undefined)
+            ) {
+                console.log(
+                    "WARNING :: An object with type='.blank' or '.blank-frame' exists with missing dimension information " +
+                        "(either the width or the height is missing). This object will be omitted in the memory model" +
+                        " diagram."
+                );
+            } else if (frame_types.includes(item.type)) {
+                stackFrames.push(item);
+            } else {
+                otherItems.push(item);
+            }
+        }
+
+        return { stack_frames: stackFrames, other_items: otherItems };
+    }
+
+    /**
+     * Return the stack-frames with generated x and y coordinates, as well as the minimum required
+     * height for drawing the stack-frames. The returned collection of stack-frames is the augmented version
+     * of the input such that the x and y coordinates of the stack-frames are determined automatically.
+     *
+     * @param stack_frames - The list of stack-frames that will be drawn
+     * (without the specified x and y coordinates)
+     * @returns - Returns the object consisting of three attributes as follows: stack-frames which will be drawn,
+     * the minimum required height of the canvas for drawing stack frames and required width for drawing all the stack
+     * frames. Notably, the last two attributes will be useful in terms of dynamically deciding the width and the height
+     * of the canvas.
+     */
+    private setStackFrameCoordinates(
+        stack_frames: DrawnEntityWithDimensions[]
+    ): {
+        StackFrames: DrawnEntityStrict[];
+        requiredHeight: number;
+        requiredWidth: number;
+    } {
+        for (const req_prop of REQUIRED_DISPLAY_PROPERTIES) {
+            if (!Object.prototype.hasOwnProperty.call(this, req_prop)) {
+                (this[req_prop] as number | undefined) = config.obj_x_padding;
+            }
+        }
+
+        let min_required_height = this.top_margin ?? config.canvas_padding!;
+        let required_width = 0;
+        const draw_stack_frames: DrawnEntityStrict[] = [];
+
+        for (const stack_frame of stack_frames) {
+            const width: number = stack_frame.width;
+            const height: number = stack_frame.height;
+
+            if (width > required_width) {
+                required_width = width;
+            }
+
+            if (stack_frame.type !== ".blank-frame") {
+                if (stack_frame.x === undefined) {
+                    stack_frame.x = this.left_margin;
+                }
+                if (stack_frame.y === undefined) {
+                    stack_frame.y = min_required_height;
+                }
+                draw_stack_frames.push(stack_frame as DrawnEntityStrict);
+            }
+
+            min_required_height = height + min_required_height;
+        }
+        if (this.left_margin !== undefined) {
```

Comment:
Overall, there's some inconsistency with some of these attributes and whether we allow them to be `undefined` or not. For example, `this.left_margin` is also used on Line 1565 without an `undefined` check.

I think the best approach is to ensure that all of these attributes are set to non-`undefined` values in the constructor, so that they are guaranteed to be `number`s` (or whatever the appropriate type is). There should be default values specified for these somewhere in the codebase already, so you should be able to use those.

Note: I'm only making this comment for `this.left_margin`, but you can probably simplify uses of other attributes in the same way.